### PR TITLE
Batch Eth1 RPC Calls

### PR DESCRIPTION
**Pull request by nisdas**
- [x] Batch JSON-RPC Calls to request a large amount of headers
- [x] Do not use the block cache to retrieve eth1 blocks.
- [x] Utilize an in-memory map to provide access for eth1 headers.
- [x] Check Chainstart for all eth1 blocks.

### CODE REVIEW

File: beacon-chain/powchain/log_processing.go [hunk 1]
```diff
@@ -259,12 +260,42 @@ func (s *Service) processPastLogs(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
+	// to store all blocks
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]:
```suggestion
	// To store all blocks.
```

File: beacon-chain/powchain/log_processing.go [hunk 2]
```diff
@@ -259,12 +260,42 @@ func (s *Service) processPastLogs(ctx context.Context) error {
...
+	headersMap := make(map[uint64]*gethTypes.Header)
+
+	// batch request the desired headers and store them in a
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]:
```suggestion
	// Batch request the desired headers and store them in a
```

File: beacon-chain/powchain/log_processing.go [hunk 3]
```diff
@@ -259,12 +260,42 @@ func (s *Service) processPastLogs(ctx context.Context) error {
...
+	// map for quick access
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]:
```suggestion
	// map for quick access.
```

File: beacon-chain/powchain/log_processing.go [hunk 4]
```diff
@@ -259,12 +260,42 @@ func (s *Service) processPastLogs(ctx context.Context) error {
...
-				if err := s.checkForChainStart(ctx, big.NewInt(int64(currentBlockNum))); err != nil {
-					return err
+			for i := currentBlockNum; i <= log.BlockNumber-1; i++ {
+				if !s.chainStartData.Chainstarted {
+					h, ok := headersMap[i]
+					if !ok {
+						if err := requestHeaders(i, i+eth1HeaderReqLimit); err != nil {
+							return err
+						}
+						// retry this block
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]:
```suggestion
						// Retry this block.
```

File: beacon-chain/powchain/log_processing.go [hunk 5]
```diff
@@ -352,24 +382,32 @@ func (s *Service) processBlksInRange(ctx context.Context, startBlk uint64, endBl
 	return nil
 }
 
-// checkForChainStart checks the given  block number for if chainstart has occurred.
-func (s *Service) checkForChainStart(ctx context.Context, blkNum *big.Int) error {
-	blk, err := s.blockFetcher.BlockByNumber(ctx, blkNum)
+// checkBlockNumberForChainStart checks the given  block number for if chainstart has occurred.
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]:
```suggestion
// checkBlockNumberForChainStart checks the given block number for if chainstart has occurred.
```

File: beacon-chain/powchain/service.go [hunk 6]
```diff
@@ -448,6 +456,43 @@ func (s *Service) processSubscribedHeaders(header *gethTypes.Header) {
 	}
 }
 
+func (s *Service) batchRequestHeaders(startBlock uint64, endBlock uint64) ([]*gethTypes.Header, error) {
```
Comment (terencechain) [inline beacon-chain/powchain/service.go]: can you add a comment here?

File: beacon-chain/powchain/testing/mock.go [hunk 7]
```diff
@@ -6,6 +6,11 @@ import (
 	"math/big"
 	"time"
 
+	"github.com/ethereum/go-ethereum/accounts/abi/bind/backends"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+
```
Comment (terencechain) [inline beacon-chain/powchain/testing/mock.go]: space
```suggestion
```

File: beacon-chain/powchain/log_processing.go [hunk 8]
```diff
@@ -259,12 +260,42 @@ func (s *Service) processPastLogs(ctx context.Context) error {
...
+	// batch request the desired headers and store them in a
+	// map for quick access
+	requestHeaders := func(startBlk uint64, endBlk uint64) error {
+		headers, err := s.batchRequestHeaders(startBlk, endBlk)
+		if err != nil {
+			return err
+		}
+		for _, h := range headers {
+			if h != nil && h.Number != nil {
+				headersMap[h.Number.Uint64()] = h
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]: Do we not need mutex protection for `headersMap`?

Comment (nisdas) [reply beacon-chain/powchain/log_processing.go]: this is all single-threaded, so using a mutex isnt required since there isnt any concurrent access

Comment (terencechain) [reply beacon-chain/powchain/log_processing.go]: Cool. What i thought but still wanted to confirm

File: beacon-chain/powchain/service.go (hunk 6 above)
Comment (nisdas) [reply beacon-chain/powchain/service.go]: added it in

**Pull request merged**